File: scripts/patchers/kernel_jb_patch_kcall10.py
"""Mixin: KernelJBPatchKcall10Mixin."""

from .kernel_jb_base import asm, _rd32, _rd64, RET, NOP, struct

# Max sysent entries in XNU (dispatch clamps at 0x22E = 558).
_SYSENT_MAX_ENTRIES = 558
# Each sysent entry is 24 bytes.
_SYSENT_ENTRY_SIZE = 24
# PAC discriminator used by the syscall dispatch (MOV X17, #0xBCAD; BLRAA X8, X17).
_SYSENT_PAC_DIVERSITY = 0xBCAD


class KernelJBPatchKcall10Mixin:
    def _find_sysent_table(self, nosys_off):
        """Find the real sysent table base.

        Strategy:
        1. Find any DATA entry whose decoded pointer == _nosys.
        2. Scan backward in 24-byte steps to find the true table start
           (entry 0 is the indirect syscall handler, NOT _nosys).
        3. Validate each backward entry: sy_call decodes to a code range
           AND the metadata fields (narg, arg_bytes) look reasonable.

        Previous bug: the old code took the first _nosys match as entry 0,
        but _nosys first appears at entry ~428 (varies by XNU build).
        """
        # Step 1: find any _nosys-matching entry
        nosys_entry = -1
        seg_start = -1
        for seg_name, vmaddr, fileoff, filesize, _ in self.all_segments:
            if "DATA" not in seg_name:
                continue
            for off in range(fileoff, fileoff + filesize - _SYSENT_ENTRY_SIZE, 8):
                val = _rd64(self.raw, off)
                decoded = self._decode_chained_ptr(val)
                if decoded == nosys_off:
                    # Verify: next entry should also have valid sy_call
                    val2 = _rd64(self.raw, off + _SYSENT_ENTRY_SIZE)
                    decoded2 = self._decode_chained_ptr(val2)
                    if decoded2 > 0 and any(
                        s <= decoded2 < e for s, e in self.code_ranges
                    ):
                        nosys_entry = off
                        seg_start = fileoff
                        break
            if nosys_entry >= 0:
                break

        if nosys_entry < 0:
            return -1

        self._log(
            f"  [*] _nosys entry found at foff 0x{nosys_entry:X}, "
            f"scanning backward for table start"
        )

        # Step 2: scan backward to find entry 0
        base = nosys_entry
        entries_back = 0
        while base - _SYSENT_ENTRY_SIZE >= seg_start:
            if entries_back >= _SYSENT_MAX_ENTRIES:
                break
            prev = base - _SYSENT_ENTRY_SIZE
            # Check sy_call decodes to valid code
            val = _rd64(self.raw, prev)
            decoded = self._decode_chained_ptr(val)
            if decoded <= 0 or not any(
                s <= decoded < e for s, e in self.code_ranges
            ):
                break
            # Check metadata looks like a sysent entry
            narg = struct.unpack_from("<H", self.raw, prev + 20)[0]
            arg_bytes = struct.unpack_from("<H", self.raw, prev + 22)[0]
            if narg > 12 or arg_bytes > 96:
                break
            base = prev
            entries_back += 1

        self._log(
            f"  [+] sysent table base at foff 0x{base:X} "
            f"({entries_back} entries before first _nosys)"
        )
        return base

    def _encode_chained_auth_ptr(self, target_foff, next_val, diversity=0,
                                  key=0, addr_div=0):
        """Encode an arm64e kernel cache auth rebase chained fixup pointer.

        Layout (DYLD_CHAINED_PTR_64_KERNEL_CACHE):
          bits[29:0]:  target (file offset)
          bits[31:30]: cacheLevel (0)
          bits[47:32]: diversity (16 bits)
          bit[48]:     addrDiv
          bits[50:49]: key (0=IA, 1=IB, 2=DA, 3=DB)
          bits[62:51]: next (12 bits, 4-byte stride delta to next fixup)
          bit[63]:     isAuth (1)
        """
        val = (
            (target_foff & 0x3FFFFFFF)
            | ((diversity & 0xFFFF) << 32)
            | ((addr_div & 1) << 48)
            | ((key & 3) << 49)
            | ((next_val & 0xFFF) << 51)
            | (1 << 63)
        )
        return struct.pack("<Q", val)

    def _extract_chain_next(self, raw_val):
        """Extract the 'next' chain field from a raw chained fixup pointer."""
        return (raw_val >> 51) & 0xFFF

    def patch_kcall10(self):
        """Replace SYS_kas_info (syscall 439) with kcall10 shellcode.

        Anchor: find _nosys function by pattern, then search DATA segments
        for the sysent table base (backward scan from first _nosys entry).

        The sysent dispatch uses BLRAA X8, X17 with X17=0xBCAD, so all
        sy_call pointers must be PAC-signed with key=IA, diversity=0xBCAD,
        addrDiv=0.  We encode the cave pointer as a proper auth rebase
        chained fixup entry to match.
        """
        self._log("\n[JB] kcall10: syscall 439 replacement")

        # Find _nosys
        nosys_off = self._resolve_symbol("_nosys")
        if nosys_off < 0:
            nosys_off = self._find_nosys()
        if nosys_off < 0:
            self._log("  [-] _nosys not found")
            return False

        self._log(f"  [+] _nosys at 0x{nosys_off:X}")

        # Find _munge_wwwwwwww
        munge_off = self._resolve_symbol("_munge_wwwwwwww")
        if munge_off < 0:
            for sym, off in self.symbols.items():
                if "munge_wwwwwwww" in sym:
                    munge_off = off
                    break

        # Find sysent table (real base via backward scan)
        sysent_off = self._find_sysent_table(nosys_off)
        if sysent_off < 0:
            self._log("  [-] sysent table not found")
            return False

        self._log(f"  [+] sysent table at file offset 0x{sysent_off:X}")

        # Entry 439 (SYS_kas_info)
        entry_439 = sysent_off + 439 * _SYSENT_ENTRY_SIZE

        # Find code cave for kcall10 shellcode (~128 bytes = 32 instructions)
        cave = self._find_code_cave(128)
        if cave < 0:
            self._log("  [-] no code cave found")
            return False

        # Build kcall10 shellcode
        # Syscall args arrive via the saved state on the stack.
        # arg[0] = pointer to a userspace buffer with {func_ptr, arg0..arg9}.
        # We unpack, call func_ptr(arg0..arg9), write results back.
        parts = [
            asm("ldr x10, [sp, #0x40]"),         # 0
            asm("ldp x0, x1, [x10, #0]"),        # 1
            asm("ldp x2, x3, [x10, #0x10]"),     # 2
            asm("ldp x4, x5, [x10, #0x20]"),     # 3
            asm("ldp x6, x7, [x10, #0x30]"),     # 4
            asm("ldp x8, x9, [x10, #0x40]"),     # 5
            asm("ldr x10, [x10, #0x50]"),         # 6
            asm("mov x16, x0"),                   # 7
            asm("mov x0, x1"),                    # 8
            asm("mov x1, x2"),                    # 9
            asm("mov x2, x3"),                    # 10
            asm("mov x3, x4"),                    # 11
            asm("mov x4, x5"),                    # 12
            asm("mov x5, x6"),                    # 13
            asm("mov x6, x7"),                    # 14
            asm("mov x7, x8"),                    # 15
            asm("mov x8, x9"),                    # 16
            asm("mov x9, x10"),                   # 17
            asm("stp x29, x30, [sp, #-0x10]!"),  # 18
            bytes([0x00, 0x02, 0x3F, 0xD6]),      # 19: BLR x16
            asm("ldp x29, x30, [sp], #0x10"),     # 20
            asm("ldr x11, [sp, #0x40]"),          # 21
            NOP,                                   # 22
            asm("stp x0, x1, [x11, #0]"),         # 23
            asm("stp x2, x3, [x11, #0x10]"),      # 24
            asm("stp x4, x5, [x11, #0x20]"),      # 25
            asm("stp x6, x7, [x11, #0x30]"),      # 26
            asm("stp x8, x9, [x11, #0x40]"),      # 27
            asm("str x10, [x11, #0x50]"),          # 28
            asm("mov x0, #0"),                     # 29
            asm("ret"),                             # 30
            NOP,                                    # 31
        ]

        for i, part in enumerate(parts):
            self.emit(cave + i * 4, part, f"shellcode+{i * 4} [kcall10]")

        # ── Patch sysent[439] with proper chained fixup encoding ────────
        #
        # The old code wrote raw VAs (struct.pack("<Q", cave_va)) which
        # breaks the chained fixup chain and produces invalid PAC-signed
        # pointers.  We now encode as auth rebase pointers matching the
        # dispatch's BLRAA X8, X17 (X17=0xBCAD).

        # Read original raw value to preserve the chain 'next' field
        old_sy_call_raw = _rd64(self.raw, entry_439)
        call_next = self._extract_chain_next(old_sy_call_raw)

        self.emit(
            entry_439,
            self._encode_chained_auth_ptr(
                cave,
                next_val=call_next,
                diversity=_SYSENT_PAC_DIVERSITY,
                key=0,       # IA
                addr_div=0,  # fixed discriminator (not address-blended)
            ),
            f"sysent[439].sy_call = cave 0x{cave:X} "
            f"(auth rebase, div=0xBCAD, next={call_next}) [kcall10]",
        )

        if munge_off >= 0:
            old_sy_munge_raw = _rd64(self.raw, entry_439 + 8)
            munge_next = self._extract_chain_next(old_sy_munge_raw)
            self.emit(
                entry_439 + 8,
                self._encode_chained_auth_ptr(
                    munge_off,
                    next_val=munge_next,
                    diversity=_SYSENT_PAC_DIVERSITY,
                    key=0,
                    addr_div=0,
                ),
                f"sysent[439].sy_munge32 = 0x{munge_off:X} "
                f"(auth rebase, next={munge_next}) [kcall10]",
            )

        # sy_return_type = SYSCALL_RET_UINT64_T (7)
        self.emit(
            entry_439 + 16,
            struct.pack("<I", 7),
            "sysent[439].sy_return_type = 7 [kcall10]",
        )

        # sy_narg = 8, sy_arg_bytes = 0x20
        self.emit(
            entry_439 + 20,
            struct.pack("<I", 0x200008),
            "sysent[439].sy_narg=8,sy_arg_bytes=0x20 [kcall10]",
        )

        return True

File: sources/FirmwarePatcher/Kernel/JBPatches/KernelJBPatchKcall10.swift
// KernelJBPatchKcall10.swift — JB kernel patch: kcall10 ABI-correct sysent[439] cave
//
// Historical note: derived from the legacy Python firmware patcher during the Swift migration.
//
// Strategy: Replace SYS_kas_info (sysent[439]) with a cave implementing
// the kcall10 primitive:
//   uap[0] = target function pointer
//   uap[1..7] = arg0..arg6
// Returns 64-bit X0 via retval and _SYSCALL_RET_UINT64_T.
//
// The cave is a standard ARM64e function body with PACIBSP/RETAB.
// The sysent entries use arm64e chained auth-rebase fixup pointers.

import Foundation

extension KernelJBPatcher {
    // MARK: - Constants

    private static let sysent_max_entries = 558
    private static let sysent_entry_size = 24
    private static let sysent_pac_diversity: UInt32 = 0xBCAD

    // kcall10 semantics
    private static let kcall10_narg: UInt16 = 8
    private static let kcall10_arg_bytes: UInt16 = 32 // 8 * 4
    private static let kcall10_return_type: UInt32 = 7 // _SYSCALL_RET_UINT64_T
    private static let kcall10_einval: UInt32 = 22

    // MARK: - Entry Point

    /// ABI-correct kcall10 patch: install a sysent[439] cave.
    @discardableResult
    func patchKcall10() -> Bool {
        log("\n[JB] kcall10: ABI-correct sysent[439] cave")

        // 1. Find _nosys.
        guard let nosysOff = resolveSymbol("_nosys") ?? findNosys() else {
            log("  [-] _nosys not found")
            return false
        }

        // 2. Find sysent table base.
        guard let sysEntOff = findSysentTable(nosysOff: nosysOff) else {
            log("  [-] sysent table not found")
            return false
        }

        let entry439 = sysEntOff + 439 * Self.sysent_entry_size

        // 3. Find a reusable 8-arg munge32 helper.
        let (mungerTarget, _, matchCount) = findMunge32ForNarg(
            sysEntOff: sysEntOff,
            narg: Self.kcall10_narg,
            argBytes: Self.kcall10_arg_bytes
        )
        guard mungerTarget >= 0 else {
            log("  [-] no unique reusable 8-arg munge32 helper found")
            return false
        }

        // 4. Build cave and allocate.
        let caveBytes = buildKcall10Cave()
        guard let caveOff = findCodeCave(size: caveBytes.count) else {
            log("  [-] no executable code cave found for kcall10")
            return false
        }

        // 5. Read original sysent[439] chain metadata.
        guard entry439 + Self.sysent_entry_size <= buffer.count else {
            log("  [-] sysent[439] outside file")
            return false
        }
        let oldSyCallRaw = buffer.readU64(at: entry439)
        let callNext = extractChainNext(oldSyCallRaw)

        let oldMungeRaw = buffer.readU64(at: entry439 + 8)
        let mungeNext = extractChainNext(oldMungeRaw)
        let mungeDiv = extractChainDiversity(oldMungeRaw)
        let mungeAddrDiv = extractChainAddrDiv(oldMungeRaw)
        let mungeKey = extractChainKey(oldMungeRaw)

        log("  [+] sysent table at file offset 0x\(String(format: "%X", sysEntOff))")
        log("  [+] sysent[439] entry at 0x\(String(format: "%X", entry439))")
        log("  [+] reusing unique 8-arg munge32 target 0x\(String(format: "%X", mungerTarget)) (\(matchCount) matching sysent rows)")
        log("  [+] cave at 0x\(String(format: "%X", caveOff)) (0x\(String(format: "%X", caveBytes.count)) bytes)")

        // 6. Emit patches.
        emit(caveOff, caveBytes,
             patchID: "jb.kcall10.cave",
             description: "kcall10 ABI-correct cave (target + 7 args -> uint64 x0)")

        emit(entry439,
             encodeChainedAuthPtr(targetFoff: caveOff, nextVal: callNext,
                                  diversity: Self.sysent_pac_diversity, key: 0, addrDiv: 0),
             patchID: "jb.kcall10.sy_call",
             description: "sysent[439].sy_call = cave 0x\(String(format: "%X", caveOff)) (auth rebase, div=0xBCAD, next=\(callNext)) [kcall10]")

        emit(entry439 + 8,
             encodeChainedAuthPtr(targetFoff: mungerTarget, nextVal: mungeNext,
                                  diversity: mungeDiv, key: mungeKey, addrDiv: mungeAddrDiv),
             patchID: "jb.kcall10.sy_munge",
             description: "sysent[439].sy_arg_munge32 = 8-arg helper 0x\(String(format: "%X", mungerTarget)) [kcall10]")

        // sy_return_type (u32) + sy_narg (u16) + sy_arg_bytes (u16)
        var metadata = Data(count: 8)
        metadata.withUnsafeMutableBytes { ptr in
            ptr.storeBytes(of: Self.kcall10_return_type.littleEndian, toByteOffset: 0, as: UInt32.self)
            ptr.storeBytes(of: Self.kcall10_narg.littleEndian, toByteOffset: 4, as: UInt16.self)
            ptr.storeBytes(of: Self.kcall10_arg_bytes.littleEndian, toByteOffset: 6, as: UInt16.self)
        }
        emit(entry439 + 16, metadata,
             patchID: "jb.kcall10.sysent_meta",
             description: "sysent[439].sy_return_type=7,sy_narg=8,sy_arg_bytes=0x20 [kcall10]")

        return true
    }

    // MARK: - Sysent Table Finder

    /// Find the real sysent table base by locating a _nosys entry then scanning backward.
    private func findSysentTable(nosysOff: Int) -> Int? {
        var nosysEntry = -1
        var segStart = -1

        // Scan DATA segments for an entry whose decoded pointer == nosysOff
        for seg in segments {
            guard seg.name.contains("DATA") else { continue }
            let sStart = Int(seg.fileOffset)
            let sEnd = sStart + Int(seg.fileSize)
            var off = sStart
            while off + Self.sysent_entry_size <= sEnd {
                let val = buffer.readU64(at: off)
                let decoded = decodeChainedPtr(val)
                if decoded == nosysOff {
                    // Confirm: next entry also decodes to a code-range address
                    let val2 = buffer.readU64(at: off + Self.sysent_entry_size)
                    let dec2 = decodeChainedPtr(val2)
                    let inCode = dec2 > 0 && codeRanges.contains { dec2 >= $0.start && dec2 < $0.end }
                    if inCode {
                        nosysEntry = off
                        segStart = sStart
                        break
                    }
                }
                off += 8
            }
            if nosysEntry >= 0 { break }
        }
        guard nosysEntry >= 0 else { return nil }

        log("  [*] _nosys entry found at foff 0x\(String(format: "%X", nosysEntry)), scanning backward for table start")

        // Scan backward in sysent_entry_size steps to find table base
        var base = nosysEntry
        var entriesBack = 0
        while base - Self.sysent_entry_size >= segStart {
            guard entriesBack < Self.sysent_max_entries else { break }
            let prev = base - Self.sysent_entry_size
            let val = buffer.readU64(at: prev)
            let decoded = decodeChainedPtr(val)
            guard decoded > 0 else { break }
            let inCode = codeRanges.contains { decoded >= $0.start && decoded < $0.end }
            guard inCode else { break }
            // Check narg and arg_bytes for sanity
            let narg = buffer.data.loadLE(UInt16.self, at: prev + 20)
            let argBytes = buffer.data.loadLE(UInt16.self, at: prev + 22)
            guard narg <= 12, argBytes <= 96 else { break }
            base = prev
            entriesBack += 1
        }

        log("  [+] sysent table base at foff 0x\(String(format: "%X", base)) (\(entriesBack) entries before first _nosys)")
        return base
    }

    // MARK: - Munger Finder

    /// Find a reusable 8-arg munge32 helper with matching metadata.
    /// Returns (targetFoff, exemplarEntry, matchCount) or (-1, -1, 0).
    private func findMunge32ForNarg(
        sysEntOff: Int,
        narg: UInt16,
        argBytes: UInt16
    ) -> (Int, Int, Int) {
        var candidates: [Int: [Int]] = [:]
        for idx in 0 ..< Self.sysent_max_entries {
            let entry = sysEntOff + idx * Self.sysent_entry_size
            guard entry + Self.sysent_entry_size <= buffer.count else { break }
            let curNarg = buffer.data.loadLE(UInt16.self, at: entry + 20)
            let curArgBytes = buffer.data.loadLE(UInt16.self, at: entry + 22)
            guard curNarg == narg, curArgBytes == argBytes else { continue }
            let rawMunge = buffer.readU64(at: entry + 8)
            let target = decodeChainedPtr(rawMunge)
            guard target > 0 else { continue }
            candidates[target, default: []].append(entry)
        }
        guard !candidates.isEmpty else { return (-1, -1, 0) }
        guard candidates.count == 1 else {
            log("  [-] multiple distinct 8-arg munge32 helpers found: " +
                candidates.keys.sorted().map { "0x\(String(format: "%X", $0))" }.joined(separator: ", "))
            return (-1, -1, 0)
        }
        let (target, entries) = candidates.first!
        return (target, entries[0], entries.count)
    }

    // MARK: - kcall10 Cave Builder

    /// Build the ABI-correct kcall10 function body.
    ///
    /// Contract:
    ///   x0 = proc*
    ///   x1 = &uthread->uu_arg[0]  (uap pointer)
    ///   x2 = &uthread->uu_rval[0] (retval pointer)
    ///
    /// uap layout (8 qwords):
    ///   [0] target function pointer
    ///   [1..7] arg0..arg6
    ///
    /// Returns EINVAL (22) on null uap/retval/target, else stores X0 into retval
    /// and returns 0 with _SYSCALL_RET_UINT64_T.
    private func buildKcall10Cave() -> Data {
        var code: [Data] = []

        // pacibsp
        code.append(ARM64.pacibsp)
        // sub sp, sp, #0x30
        code.append(encodeU32k(0xD100_C3FF)) // sub sp, sp, #0x30
        // stp x21, x22, [sp]
        code.append(encodeU32k(0xA900_5BF5)) // stp x21, x22, [sp]
        // stp x19, x20, [sp, #0x10]
        code.append(encodeU32k(0xA901_53F3)) // stp x19, x20, [sp, #0x10]
        // stp x29, x30, [sp, #0x20]
        code.append(encodeU32k(0xA902_7BFD)) // stp x29, x30, [sp, #0x20]
        // add x29, sp, #0x20
        code.append(encodeU32k(0x9100_83FD)) // add x29, sp, #0x20
        // mov w19, #22  (EINVAL = 22 = 0x16)
        code.append(encodeU32k(0x5280_02D3)) // movz w19, #0x16
        // mov x20, x1  (save uap)
        code.append(encodeU32k(0xAA01_03F4)) // mov x20, x1
        // mov x21, x2  (save retval)
        code.append(encodeU32k(0xAA02_03F5)) // mov x21, x2
        // cbz x20, #0x30  (null uap → skip to exit, 12 instrs forward)
        code.append(encodeU32k(0xB400_0194)) // cbz x20, #+0x30
        // cbz x21, #0x2c  (null retval → skip to exit)
        code.append(encodeU32k(0xB400_0175)) // cbz x21, #+0x2c
        // ldr x16, [x20]  (target = uap[0])
        code.append(encodeU32k(0xF940_0290)) // ldr x16, [x20]
        // cbz x16, #0x24  (null target → skip, 9 instrs)
        code.append(encodeU32k(0xB400_0130)) // cbz x16, #+0x24
        // ldp x0, x1, [x20, #0x8]
        code.append(encodeU32k(0xA940_8680)) // ldp x0, x1, [x20, #0x8]
        // ldp x2, x3, [x20, #0x18]
        code.append(encodeU32k(0xA941_8E82)) // ldp x2, x3, [x20, #0x18]
        // ldp x4, x5, [x20, #0x28]
        code.append(encodeU32k(0xA942_9684)) // ldp x4, x5, [x20, #0x28]
        // ldr x6, [x20, #0x38]
        code.append(encodeU32k(0xF940_1E86)) // ldr x6, [x20, #0x38]
        // mov x7, xzr
        code.append(encodeU32k(0xAA1F_03E7)) // mov x7, xzr
        // blr x16
        code.append(encodeU32k(0xD63F_0200)) // blr x16
        // str x0, [x21]  (store result in retval)
        code.append(encodeU32k(0xF900_02A0)) // str x0, [x21]
        // mov w19, #0
        code.append(encodeU32k(0x5280_0013)) // movz w19, #0
        // mov w0, w19  (return value)
        code.append(encodeU32k(0x2A13_03E0)) // mov w0, w19
        // ldp x21, x22, [sp]
        code.append(encodeU32k(0xA940_5BF5)) // ldp x21, x22, [sp]
        // ldp x19, x20, [sp, #0x10]
        code.append(encodeU32k(0xA941_53F3)) // ldp x19, x20, [sp, #0x10]
        // ldp x29, x30, [sp, #0x20]
        code.append(encodeU32k(0xA942_7BFD)) // ldp x29, x30, [sp, #0x20]
        // add sp, sp, #0x30
        code.append(encodeU32k(0x9100_C3FF)) // add sp, sp, #0x30
        // retab
        code.append(ARM64.retab)

        return code.reduce(Data(), +)
    }

    // MARK: - Chain Pointer Helpers

    /// Encode an arm64e kernel cache auth-rebase chained fixup pointer.
    private func encodeChainedAuthPtr(
        targetFoff: Int,
        nextVal: UInt32,
        diversity: UInt32,
        key: UInt32,
        addrDiv: UInt32
    ) -> Data {
        let val: UInt64 =
            (UInt64(targetFoff) & 0x3FFF_FFFF) |
            (UInt64(diversity & 0xFFFF) << 32) |
            (UInt64(addrDiv & 1) << 48) |
            (UInt64(key & 3) << 49) |
            (UInt64(nextVal & 0xFFF) << 51) |
            (1 << 63)
        return withUnsafeBytes(of: val.littleEndian) { Data($0) }
    }

    private func extractChainNext(_ raw: UInt64) -> UInt32 {
        UInt32((raw >> 51) & 0xFFF)
    }

    private func extractChainDiversity(_ raw: UInt64) -> UInt32 {
        UInt32((raw >> 32) & 0xFFFF)
    }

    private func extractChainAddrDiv(_ raw: UInt64) -> UInt32 {
        UInt32((raw >> 48) & 1)
    }

    private func extractChainKey(_ raw: UInt64) -> UInt32 {
        UInt32((raw >> 49) & 3)
    }

    // MARK: - Encoding Helper

    private func encodeU32k(_ value: UInt32) -> Data {
        ARM64.encodeU32(value)
    }
}
